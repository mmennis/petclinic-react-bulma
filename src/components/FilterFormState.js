import React from 'react'
import { Form, Button, Tile } from 'react-bulma-components'
import PropTypes from 'prop-types'

export default class FilterFormState extends React.Component {

    constructor(props) {
        super(props)
        this.state = {
            stateFilter: ''
        }
        this.handleChange = this.handleChange.bind(this)
        this.submitForm = this.submitForm.bind(this)
    }

    handleChange = (e) => {
        this.setState({ stateFilter: e.target.value.trim() })
    }

    submitForm = (e) => {
        e.preventDefault()
        this.props.handleFilterUpdate(this.state)
    }

    render() {
        return (
            <div>
                <form onSubmit={this.submitForm} data-testid="form">
                    <Form.Field>
                        <Tile kind="parent">
                            <Tile size={1}>
                                <Form.Label>State:</Form.Label>
                            </Tile>
                            <Tile size={3}>
                                <Form.Control>
                                    <Form.Input 
                                        name="stateFilter" 
                                        type="text" 
                                        onChange={this.handleChange} 
                                        value={this.state.stateFilter} 
                                        data-testid="input"
                                    />
                                </Form.Control>
                            </Tile>
                            <Tile size={1}>
                                <Form.Control>
                                    <Button type="submit" color="primary" onClick={this.submitForm}>Submit</Button>
                                </Form.Control>
                            </Tile>
                        </Tile>
                    </Form.Field>
                </form>
            </div>
        )
    }
}

FilterFormState.propTypes = {
    handleFilterUpdate: PropTypes.func.isRequired
}