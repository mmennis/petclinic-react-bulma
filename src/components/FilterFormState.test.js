import React from 'react'
import { create } from 'react-test-renderer'
import { render, fireEvent } from '@testing-library/react'
import FilterFormState from './FilterFormState'

describe('FilterFormState', () => {

    afterEach(() => {
        jest.clearAllMocks()
    })

    it('should render as per the snapshot', () => {
        const callback = jest.fn()
        const stateFilterForm = create(<FilterFormState handleFilterUpdate={callback}/>)
        expect(stateFilterForm.toJSON()).toMatchSnapshot()
    })

    it('should log an error to the console if required owner props missing', () => {
        console.error = jest.fn();
        React.createElement(FilterFormState)
        expect(console.error).toHaveBeenCalledTimes(1)
    })

    it('should call the handleFilterUpdate function with an expected value', () => {
        const callback = jest.fn()
        const { getByTestId } = render(<FilterFormState handleFilterUpdate={callback} />)
        fireEvent.change(getByTestId("input"), { target: { value: 'CA' }})
        fireEvent.submit(getByTestId("form"))
        expect(callback).toHaveBeenCalledWith({ stateFilter: 'CA' })
    })

    it('should ignore whitespace and pass on trimmed entry', () => {
        const callback = jest.fn()
        const { getByTestId } = render(<FilterFormState handleFilterUpdate={callback} />)
        fireEvent.change(getByTestId("input"), { target: { value: 'CA ' }})
        fireEvent.submit(getByTestId("form"))
        expect(callback).toHaveBeenCalledWith({ stateFilter: 'CA' })
    })

    it('should convert entry to upper case', () => {
        const callback = jest.fn()
        const { getByTestId } = render(<FilterFormState handleFilterUpdate={callback} />)
        fireEvent.change(getByTestId("input"), { target: { value: 'ca' }})
        fireEvent.submit(getByTestId("form"))
        expect(callback).toHaveBeenCalledWith({ stateFilter: 'CA' })
    })

})