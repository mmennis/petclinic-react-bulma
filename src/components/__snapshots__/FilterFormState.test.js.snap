// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterFormState should render as per the snapshot 1`] = `
<div>
  <form
    data-testid="form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <div
        className="tile is-parent"
      >
        <div
          className="tile is-1"
        >
          <label
            className="label"
          >
            State:
          </label>
        </div>
        <div
          className="tile is-3"
        >
          <div
            className="control"
          >
            <input
              className="input"
              data-testid="input"
              disabled={false}
              name="stateFilter"
              onChange={[Function]}
              placeholder=""
              readOnly={false}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="tile is-1"
        >
          <div
            className="control"
          >
            <button
              className="is-primary button"
              disabled={false}
              onClick={[Function]}
              tabIndex={0}
              type="submit"
            >
              Submit
            </button>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
